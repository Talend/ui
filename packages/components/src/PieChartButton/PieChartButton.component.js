import React from 'react';
import PropTypes from 'prop-types';
import { arc } from 'd3-shape';
import { Button, OverlayTrigger, Popover } from 'react-bootstrap';
import classnames from 'classnames';
import TooltipTrigger from '../TooltipTrigger';
import Skeleton from '../Skeleton';
import theme from './PieChartButton.scss';

const MIN_SIZE = 20;
const MAX_SIZE = 50;
const MAX_PERCENT = 100;
const BASE_INNER_RADIUS = 6;
const BASE_OUTER_RADIUS = 9;
const BASE_PAD_ANGLE = 0.2;
const INNER_RADIUS_PER_PIXEL = 0.4;
const OUTER_RADIUS_PER_PIXEL = 0.45;
const PAD_ANGLE_PER_PIXEL = 0.0033;

// we need just one instance of this, it's just a generator
const arcGen = arc();

const displaySizes = {
	small: 20,
	medium: 35,
	large: 50,
};

/**
 * This function return the angle for a given percentage
 * @param {number} percentage the percentage to calculate the angle
 */
export function getAngle(percentage) {
	return percentage * 2 / 100 * Math.PI;
}

/**
 * this function generates chart's empty part
 * @param {array} values the values shown in the graph
 * @param {object} size the current size
 * @param {function} arcGen the arc generator
 * @param {number} minimumPercentage the minimum percentage to be shown
 */
export function getEmptyPartCircle(values, size, minimumPercentage) {
	const allPercentages = values.reduce((acc, value) => acc + value.percentageShown, 0);
	if (allPercentages >= MAX_PERCENT - minimumPercentage) {
		return null;
	}

	const arcGenerated = arcGen({
		innerRadius: size.innerRadius,
		outerRadius: size.outerRadius,
		padAngle: size.padAngle,
		startAngle: getAngle(allPercentages),
		endAngle: Math.PI * 2,
	});
	return (
		<path
			className={classnames(theme['tc-pie-chart-color-alto'], 'tc-pie-chart-color-alto')}
			d={arcGenerated}
			transform={`translate(${size.svgSize / 2},${size.svgSize / 2})`}
		/>
	);
}

/**
 * This function return the sum of the percentages below the given index
 * @param {array} values the values to get when we have to start
 * @param {number} index the current index
 */
export function getPercentageToIndex(values, index) {
	return values.reduce((acc, value, i) => {
		if (i < index) {
			return acc + value.percentageShown;
		}
		return acc;
	}, 0);
}

/**
 * This function generate a part of circle for a value
 * @param {object} value the value to generate
 * @param {number} index the current index
 * @param {array} values the values to get when we have to start
 * @param {object} size the current graph size
 */
export function getCircle(value, index, values, size) {
	const percentagesDone = getPercentageToIndex(values, index);

	const arcGenerated = arcGen({
		innerRadius: size.innerRadius,
		outerRadius: size.outerRadius,
		padAngle: size.padAngle,
		startAngle: getAngle(percentagesDone),
		endAngle: getAngle(percentagesDone + value.percentageShown),
	});

	return (
		<path
			key={index}
			d={arcGenerated}
			className={classnames(
				theme[`tc-pie-chart-color-${value.color}`],
				`tc-pie-chart-color-${value.color}`,
			)}
			transform={`translate(${size.svgSize / 2},${size.svgSize / 2})`}
		/>
	);
}

/**
 * This function is just a sort
 * @param {object} a an element with a percentageShown to compare
 * @param {object} b an other element with a percentageShown to compare
 */
function sortElements(a, b) {
	if (a.percentageShown > b.percentageShown) {
		return -1;
	}
	if (a.percentageShown < b.percentageShown) {
		return 1;
	}
	return 0;
}

/**
 * It increase values below minPercentage & decrease values above if some are increased
 * setMinimumPercentage([{percentage: 50}, {percentage: 2}], 5) would result on
 * [{percentage: 50, percentageShown: 47}, {percentage: 2, percentageShown: 5}]
 * @param {*} values the set of values
 * @param {*} minimumPercentage the minimum value we have to show
 * @param {*} amountToSubtract the amount to decrease
 */
export function distributePercentages(values, minimumPercentage, amountToSubtract) {
	/**
	 * This function decrease the percentage shown & the amount to subtract by 1
	 * @param {object} element the current element
	 */
	function decreaseElement(element) {
		if (amountToSubtract > 0) {
			// eslint-disable-next-line no-param-reassign
			element.percentageShown -= 1;
			// eslint-disable-next-line no-param-reassign
			amountToSubtract -= 1;
		}
	}

	while (amountToSubtract > 0) {
		const elementsToDecrease = values
			.filter(value => value.percentageShown > minimumPercentage)
			.sort(sortElements);
		if (elementsToDecrease.length > 0) {
			elementsToDecrease.forEach(decreaseElement);
		} else {
			// eslint-disable-next-line no-param-reassign
			amountToSubtract = 0;
		}
	}

	return values;
}

/**
 * This function sets minimum percentage show with the minimum percentage shown
 * @param {array} values the set of values
 * @param {number} minimumPercentage the minimum value we have to show
 */
export function setMinimumPercentage(model, minimumPercentage) {
	if (!model) {
		return [];
	}

	let amountToSubtract = 0;

	const valuesMins = model.map(value => {
		if (value.percentage && value.percentage < minimumPercentage) {
			amountToSubtract += minimumPercentage - value.percentage;
			return { ...value, percentageShown: minimumPercentage };
		}
		return { ...value, percentageShown: value.percentage };
	});

	return distributePercentages(valuesMins, minimumPercentage, amountToSubtract);
}

/**
 * This function add an OverlayTrigger wrapping the button if defined
 * @param {Element} btn the current button
 * @param {string} overlayPlacement the overlay placement
 * @param {Element} overlayComponent the overlay component
 * @param {string} overlayId the id to be set for the overlay
 */
export function decorateWithOverlay(btn, overlayPlacement, overlayComponent, overlayId) {
	if (!overlayComponent) {
		return btn;
	}

	return (
		<span>
			<OverlayTrigger
				trigger="click"
				rootClose
				placement={overlayPlacement}
				overlay={<Popover id={overlayId}>{overlayComponent}</Popover>}
			>
				{btn}
			</OverlayTrigger>
		</span>
	);
}

/**
 * This function wrap the button with a TooltipTrigger
 * @param {Element} btn the button element ( may be wrapped by overlay trigger )
 * @param {boolean} tooltip tell if the tooltip has to be showed
 * @param {string} label the label to show on the tooltip
 * @param {string} tooltipPlacement the tooltip placement
 */
export function decorateWithTooltip(btn, tooltip, label, tooltipPlacement) {
	if (!tooltip || !label) {
		return btn;
	}
	return (
		<TooltipTrigger label={label} tooltipPlacement={tooltipPlacement}>
			{btn}
		</TooltipTrigger>
	);
}

/**
 * This function wrap the event when we don't have an overlay component
 * @param {function} mouseEvent the event to wrap ( mouseClick )
 * @param {Element} overlayComponent tell if there is an overlay component
 * @param {string} label the label of the component
 * @param {object} rest the rest of the props
 * @param {object} model the model of the component
 */
export function wrapMouseEvent(mouseEvent, overlayComponent, label, rest, model) {
	if (overlayComponent || !mouseEvent) {
		return null;
	}
	return event =>
		mouseEvent(event, {
			action: { label, ...rest },
			model,
		});
}

/**
 * This function return useful stuff to build the graph of the loader
 * @param {number} size the size in px of the graph
 * @param {string} display the chosen display if given
 */
export function getDisplaySize(size, display) {
	let currentSize = size;
	if (!currentSize && display) {
		currentSize = displaySizes[display];
	}

	const pixelNumber = currentSize - MIN_SIZE;
	return {
		svgSize: currentSize,
		innerRadius: parseInt(BASE_INNER_RADIUS + INNER_RADIUS_PER_PIXEL * pixelNumber, 10),
		outerRadius: parseInt(BASE_OUTER_RADIUS + OUTER_RADIUS_PER_PIXEL * pixelNumber, 10),
		padAngle: BASE_PAD_ANGLE - PAD_ANGLE_PER_PIXEL * pixelNumber,
	};
}

/**
 * This function check if the type is a number & check with min/max
 * @param {object} props list of component props
 * @param {string} propName current prop name
 * @param {string} componentName component name
 */
function propTypeCheckSize(props, propName, componentName) {
	if (props[propName] == null) {
		return null;
	}
	if (typeof props[propName] !== 'number') {
		return new Error(
			`Invalid type of ${propName} supplied to ${componentName} : ${typeof props[
				propName
			]}. Validation failed.`,
		);
	} else if (props[propName] < MIN_SIZE || props[propName] > MAX_SIZE) {
		return new Error(
			`Invalid prop ${propName} supplied to ${componentName} : ${
				props[propName]
			}. Validation failed.`,
		);
	}
	return null;
}

/**
 * This function return the showed value on the chart
 * @param {array} model the pie chart model
 * @param {index} index current index showed
 */
function getShowedValue(model, index) {
	if (!model) {
		return {};
	}
	return model[index];
}

function PieChartButton({
	model,
	labelIndex,
	className,
	loading,
	minimumPercentage,
	display,
	label,
	overlayComponent,
	overlayPlacement,
	overlayId,
	onClick,
	onMouseDown,
	hideLabel,
	size,
	tooltip,
	tooltipPlacement,
	...rest
}) {
	const sizeObject = getDisplaySize(size, display);

	if (loading) {
		return (
			<Button
				className={classnames(theme['tc-pie-chart-loading'], 'tc-pie-chart-loading', {
					[theme['tc-pie-chart-loading-no-label']]: hideLabel,
					'tc-pie-chart-loading-no-label': hideLabel,
				})}
			>
				<Skeleton
					type={Skeleton.TYPES.circle}
					width={sizeObject.svgSize}
					height={sizeObject.svgSize}
					className={classnames(
						theme['tc-pie-chart-loading-circle'],
						'tc-pie-chart-loading-circle',
					)}
				/>
				{!hideLabel && <Skeleton type={Skeleton.TYPES.text} size={Skeleton.SIZES.small} />}
			</Button>
		);
	}

	const labelValue = getShowedValue(model, labelIndex);
	const preparedValues = setMinimumPercentage(model, minimumPercentage);
	const rClick = wrapMouseEvent(onClick, overlayComponent, label, rest, model);
	const rMouseDown = wrapMouseEvent(onMouseDown, overlayComponent, label, rest, model);

	let btn = (
		<Button
			className={classnames(theme['tc-pie-chart'], 'tc-pie-chart', className)}
			onMouseDown={rMouseDown}
			onClick={rClick}
			{...rest}
		>
			<svg
				width={sizeObject.svgSize}
				height={sizeObject.svgSize}
				className={classnames(theme['tc-pie-chart-graph'], 'tc-pie-chart-graph')}
			>
				{preparedValues.map((value, index) => getCircle(value, index, preparedValues, sizeObject))}
				{getEmptyPartCircle(preparedValues, sizeObject, minimumPercentage)}
			</svg>
			<div
				className={classnames(
					theme['tc-pie-chart-label'],
					'tc-pie-chart-label',
					theme[`tc-pie-chart-color-${labelValue.color}`],
					`tc-pie-chart-color-${labelValue.color}`,
				)}
			>
				{!hideLabel && labelValue.percentage && `${labelValue.percentage} %`}
			</div>
		</Button>
	);

	btn = decorateWithOverlay(btn, overlayPlacement, overlayComponent, overlayId);
	btn = decorateWithTooltip(btn, tooltip, label, tooltipPlacement);

	return btn;
}

PieChartButton.propTypes = {
	className: PropTypes.string,
	display: PropTypes.oneOf(['small', 'medium', 'large']),
	loading: PropTypes.bool,
	hideLabel: PropTypes.bool,
	label: PropTypes.string,
	labelIndex: PropTypes.number,
	getComponent: PropTypes.func,
	minimumPercentage: PropTypes.number.isRequired,
	model: PropTypes.arrayOf(
		PropTypes.shape({
			color: PropTypes.oneOf([
				'rio-grande',
				'chestnut-rose',
				'lightning-yellow',
				'slate-gray',
				'silver-chalice',
			]),
			percentage: PropTypes.number.isRequired,
		}).isRequired,
	),
	onClick: PropTypes.func,
	onMouseDown: PropTypes.func,
	overlayComponent: PropTypes.element,
	overlayId: PropTypes.string,
	overlayPlacement: OverlayTrigger.propTypes.placement,
	size: propTypeCheckSize,
	tooltip: PropTypes.bool,
	tooltipPlacement: OverlayTrigger.propTypes.placement,
};

PieChartButton.defaultProps = {
	labelIndex: 0,
	minimumPercentage: 5,
	display: 'small',
	tooltipPlacement: 'top',
	overlayPlacement: 'bottom',
	overlayId: 'pie-chart-popover',
};

PieChartButton.displayName = 'PieChartButton';

export default PieChartButton;
