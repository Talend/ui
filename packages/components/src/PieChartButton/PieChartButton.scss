@import '~@talend/bootstrap-theme/src/theme/animation';
$tc-pie-chart-font-size: $font-size-large !default;
$tc-pie-chart-loading-min-height-no-label: 20px !default;
$tc-pie-chart-loading-min-height: 25px !default;
$tc-pie-chart-skeleton-label-width: 40px !default;
$tc-pie-chart-skeleton-label-height: 14px !default;
$tc-pie-chart-loading-margin-left: 4px !default;
$tc-pie-chart-hover-color: $scooter !default;

$tc-pie-chart-colors: (
	rio-grande $rio-grande,
	chestnut-rose $chestnut-rose,
	lightning-yellow $lightning-yellow,
	slate-gray $slate-gray,
	silver-chalice $silver-chalice,
	alto $alto
) !default;

@mixin apply-color($color) {
	color: $color;
	stroke: $color;
	fill: $color;
}

.tc-pie-chart-loading {
	display: flex;
	align-items: center;
	min-height: $tc-pie-chart-loading-min-height;

	&.tc-pie-chart-loading-no-label {
		min-height: $tc-pie-chart-loading-min-height-no-label;
	}

	&-circle {
		// adjust to be like loaded component
		margin-left: $tc-pie-chart-loading-margin-left;
		margin-right: $padding-smaller;
	}
}

// PieChart styles
.tc-pie-chart,
.tc-pie-chart-loading {
	font-size: $tc-pie-chart-font-size;
	display: flex;
	align-items: center;

	background-color: transparent;
	border: none;
	color: $gray;

	box-shadow: none;
	line-height: unset;
	padding: 0;

	@each $type in $tc-pie-chart-colors {
		.tc-pie-chart-color-#{nth($type, 1)} {
			@include apply-color(#{nth($type, 2)});
		}
	}

	&:hover {
		@each $type in $tc-pie-chart-colors {
			.tc-pie-chart-color-#{nth($type, 1)} {
				@include apply-color($tc-pie-chart-hover-color);
			}
		}
	}

	& &-graph {
		margin-right: $padding-smaller;
		height: inherit;
		width: inherit;
	}

	&:hover,
	&:active,
	&:focus {
		box-shadow: none;
		background-color: transparent;
		border-color: $scooter;
		color: $scooter;
	}

	&.active {
		background-color: $brand-primary;
		border-color: $brand-primary;
		color: $white;

		&:hover,
		&:active,
		&:focus {
			background-color: $scooter;
			border-color: $scooter;
			color: $white;
		}
	}
}
