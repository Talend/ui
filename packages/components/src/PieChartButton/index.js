import PieChartButton from './PieChartButton.component';

export default PieChartButton;
