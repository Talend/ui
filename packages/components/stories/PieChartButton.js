import React from 'react';
import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';

import ActionButton from '../src/Actions/ActionButton';
import PieChartButton from '../src/PieChartButton';

const pieChartData1 = [
	{
		color: 'rio-grande',
		percentage: 50,
	},
	{
		color: 'chestnut-rose',
		percentage: 12,
	},
	{
		color: 'lightning-yellow',
		percentage: 1,
	},
	{
		color: 'slate-gray',
		percentage: 4,
	},
	{
		color: 'silver-chalice',
		percentage: 3,
	},
];

const pieChartData2 = [
	{
		color: 'rio-grande',
		percentage: 15,
	},
	{
		color: 'chestnut-rose',
		percentage: 15,
	},
	{
		color: 'lightning-yellow',
		percentage: 60,
	},
	{
		color: 'slate-gray',
		percentage: 2,
	},
];

const sizes = [20, 22, 25, 30, 35, 40, 45, 50];

const onClick = action('You clicked me');
const onMouseDown = action('You mousedown me');
const overlayComponent = <div>I am an overlay</div>;

const stories = storiesOf('PieChartButton', module);
if (!stories.addWithInfo) {
	stories.addWithInfo = stories.add;
}

stories
	.addDecorator(story => <div className="col-lg-offset-2 col-lg-8">{story()}</div>)
	.addWithInfo('default', () => (
		<div>
			<p>Small :</p>
			<PieChartButton display="small" model={pieChartData1} onClick={onClick} />
			<p>custom size</p>
			<PieChartButton size={25} model={pieChartData1} onClick={onClick} />
			<p>Medium :</p>
			<PieChartButton display="medium" model={pieChartData1} onMouseDown={onMouseDown} />
			<p>Large : </p>
			<PieChartButton display="large" model={pieChartData1} />
			<p>with other data :</p>
			<PieChartButton display="medium" model={pieChartData2} />
			<p>with without label :</p>
			<PieChartButton display="medium" model={pieChartData2} hideLabel />
			<p>with overlay component</p>
			<PieChartButton
				display="medium"
				labelIndex={2}
				model={pieChartData2}
				overlayComponent={overlayComponent}
				overlayId="id-popover"
				onClick={onClick}
			/>
			<p>with a tooltip</p>
			<PieChartButton
				display="medium"
				model={pieChartData2}
				label="this is a tooltip"
				tooltip
				tooltipPlacement="right"
				onClick={onClick}
			/>
		</div>
	))
	.addWithInfo('lot of sizes', () => (
		<div>
			{sizes.map(size => (
				<div key={size}>
					<p>{size}px</p>
					<PieChartButton size={size} model={pieChartData1} />
				</div>
			))}
		</div>
	))
	.addWithInfo('lot of loading', () => (
		<div>
			<div>
				{sizes.map(size => (
					<div key={size}>
						<p>{size}px</p>
						<PieChartButton loading size={size} model={pieChartData1} />
					</div>
				))}
			</div>
		</div>
	))
	.addWithInfo('Loading state', () => {
		class WithLayout extends React.Component {
			constructor() {
				super();
				this.changeState = this.changeState.bind(this);
				this.state = { loading: true };
			}

			changeState() {
				this.setState(prevState => ({
					loading: !prevState.loading,
				}));
			}

			render() {
				return (
					<div>
						<ActionButton label="changestatus" onClick={this.changeState} />
						<p>Small :</p>
						<PieChartButton display="small" model={pieChartData1} loading={this.state.loading} />
						<p>Medium :</p>
						<PieChartButton display="medium" model={pieChartData1} loading={this.state.loading} />
						<p>Large :</p>
						<PieChartButton display="large" model={pieChartData1} loading={this.state.loading} />
						<p>Small without label:</p>
						<PieChartButton
							display="small"
							hideLabel
							model={pieChartData1}
							loading={this.state.loading}
						/>
						<p>Medium without label:</p>
						<PieChartButton
							display="medium"
							hideLabel
							model={pieChartData1}
							loading={this.state.loading}
						/>
						<p>Large without label:</p>
						<PieChartButton
							display="large"
							hideLabel
							model={pieChartData1}
							loading={this.state.loading}
						/>
						<p>after large</p>
					</div>
				);
			}
		}
		return <WithLayout />;
	});
