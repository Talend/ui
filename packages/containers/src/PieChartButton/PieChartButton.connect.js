import React from 'react';
import PropTypes from 'prop-types';
import Immutable from 'immutable';
import omit from 'lodash/omit';
import { componentState, cmfConnect, Inject } from '@talend/react-cmf';
import PieChartButton from '@talend/react-components/lib/PieChartButton';

export const DEFAULT_STATE = new Immutable.Map({});

export function ContainerPieChartButton(props) {
	let overlayComponent = null;
	let onClick = null;

	if (typeof props.overlayComponent === 'string' && props.overlayComponent) {
		overlayComponent = (
			<Inject component={props.overlayComponent} {...props.overlayComponentProps} />
		);
	}

	if (!props.onClick) {
		onClick = (event, data) => {
			if (props.actionCreator) {
				props.dispatchActionCreator(props.actionCreator, event, data);
			} else {
				props.dispatch({
					model: props.model,
					...props.payload,
				});
			}
		};
	}

	const state = props.state || DEFAULT_STATE;
	const newProps = {
		...omit(props, cmfConnect.INJECTED_PROPS.concat(['getComponent', 'initialState'])),
		model: state.get('model', props.model),
		inProgress: state.get('inProgress', props.inProgress),
		overlayComponent,
		onClick,
	};

	return <PieChartButton {...newProps} />;
}

ContainerPieChartButton.displayName = 'Container(PieChartButton)';

ContainerPieChartButton.propTypes = {
	...componentState.propTypes,
	actionCreator: PropTypes.string,
	dispatch: PropTypes.func,
	dispatchActionCreator: PropTypes.func,
	model: PropTypes.object,
	onClick: PropTypes.func,
	overlayComponent: PropTypes.string,
	overlayComponentProps: PropTypes.object,
	payload: PropTypes.object,
};

export default cmfConnect({
	componentId: ownProps => ownProps.componentId || ownProps.id,
	defaultState: DEFAULT_STATE,
})(ContainerPieChartButton);
