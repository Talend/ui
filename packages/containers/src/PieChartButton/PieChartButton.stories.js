import React from 'react';
import PieChartButton from '.';

const payload = { type: 'CLICK' };

const initialState1 = {
	model: [{ percentage: 10, color: 'rio-grande' }],
};
const initialState2 = {
	model: [
		{ percentage: 10, color: 'rio-grande' },
		{ percentage: 15, color: 'chestnut-rose' },
		{ percentage: 5, color: 'lightning-yellow' },
		{ percentage: 20, color: 'dove-gray' },
	],
};
const initialStateInProgress = {
	loading: true,
};

export default {
	title: 'PieChartButton',
};

export const Default = () => (
	<div>
		<p>Loading :</p>
		<PieChartButton id="pieChart1" initialState={initialStateInProgress} payload={payload} />
		<p>With Data :</p>
		<PieChartButton id="pieChart2" initialState={initialState1} payload={payload} />
		<p>With Other data :</p>
		<PieChartButton id="pieChart3" initialState={initialState2} display="large" payload={payload} />
	</div>
);
