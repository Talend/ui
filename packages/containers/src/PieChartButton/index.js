import PieChartButton from './PieChartButton.connect';

export default PieChartButton;
