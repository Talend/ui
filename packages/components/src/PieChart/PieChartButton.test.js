import React from 'react';
import { shallow, mount } from 'enzyme';
import PieChartButton, { decorateWithOverlay, wrapMouseEvent } from './PieChartButton.component';

describe('PieChartButton', () => {
	describe('snapshots render', () => {
		const pieChartData = [
			{
				color: 'rio-grande',
				percentage: 50,
			},
			{
				color: 'chestnut-rose',
				percentage: 12,
			},
			{
				color: 'lightning-yellow',
				percentage: 1,
			},
			{
				color: 'dove-gray',
				percentage: 4,
			},
			{
				color: 'silver-chalice',
				percentage: 3,
			},
		];
		it('should render a PieChartButton', () => {
			const wrapper = shallow(<PieChartButton display="small" model={pieChartData} />);
			expect(wrapper.getElement()).toMatchSnapshot();
		});
		it('should render nothing', () => {
			const wrapper = shallow(
				<PieChartButton available={false} display="medium" labelIndex={2} model={pieChartData} />,
			);

			expect(wrapper.getElement()).toBeNull();
		});
		it('should trigger onClick', () => {
			const onClick = jest.fn();
			const event = {};
			const wrapper = shallow(
				<PieChartButton label="my label" display="small" model={pieChartData} onClick={onClick} />,
			);

			wrapper
				.find('Button')
				.at(0)
				.simulate('click', event);

			expect(onClick).toHaveBeenCalledWith(event, {
				action: {
					label: 'my label',
				},
				model: pieChartData,
			});
		});

		it('should render a PieChartButton with an overlay', () => {
			const overlayComponent = <div>I am an overlay</div>;
			const wrapper = shallow(
				<PieChartButton
					display="medium"
					labelIndex={2}
					model={pieChartData}
					overlayComponent={overlayComponent}
					overlayId="id-popover"
				/>,
			);

			expect(wrapper.getElement()).toMatchSnapshot();
		});
		it('should called refs methods', () => {
			// given
			const overlayComponent = <div className="fake-overlay" />;
			// when
			const myButtonRef = jest.fn();
			const myOverlayRef = jest.fn();
			mount(
				<PieChartButton
					display="medium"
					labelIndex={2}
					model={pieChartData}
					overlayComponent={overlayComponent}
					overlayId="id-popover"
					buttonRef={myButtonRef}
					overlayRef={myOverlayRef}
				/>,
			);
			expect(myButtonRef).toHaveBeenCalled();
			expect(myOverlayRef).toHaveBeenCalled();
		});
	});
});

describe('decorateWithOverlay', () => {
	it('should return the same component if no overlayComponent', () => {
		// given
		const btn = <div className="fake-button-element" />;
		// when
		const modified = decorateWithOverlay(btn);
		// then
		expect(modified).toBe(btn);
	});

	it('should return the component wrapped', () => {
		// given
		const btn = <div className="fake-button-element" />;
		const overlayComponent = <div className="fake-overlay" />;
		// when
		const modified = decorateWithOverlay(btn, 'top', overlayComponent, 'id-test');
		// then
		expect(modified).not.toBe(btn);
		expect(modified).toMatchSnapshot();
	});
	it('should trigger ref function', () => {
		// given
		const btn = <div className="fake-button-element" />;
		const overlayComponent = <div className="fake-overlay" />;
		// when
		const myBindRef = jest.fn();
		function OverlayCmp() {
			return (
				<div>{decorateWithOverlay(btn, 'top', overlayComponent, 'myDumbOverlay', myBindRef)}</div>
			);
		}
		mount(<OverlayCmp />);
		expect(myBindRef).toHaveBeenCalled();
	});
});

describe('wrapMouseEvent', () => {
	it('should return null if there is an overlay component', () => {
		// given
		const onClick = jest.fn();
		const overlayComponent = <div className="fake-overlay" />;

		// when
		const result = wrapMouseEvent(onClick, overlayComponent, 'label');

		// then
		expect(result).toBeNull();
	});
	it('should return a wrapped handler if there is no overlay component', () => {
		// given
		const onClick = jest.fn();

		// when
		const result = wrapMouseEvent(onClick, null, 'label');

		// then
		expect(result).not.toBeNull();
		expect(onClick).not.toHaveBeenCalled();
		result();
		expect(onClick).toHaveBeenCalled();
	});
});
