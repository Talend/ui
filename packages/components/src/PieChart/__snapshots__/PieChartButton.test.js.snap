// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PieChartButton snapshots render should render a PieChartButton 1`] = `
<Button
  active={false}
  block={false}
  bsClass="btn"
  bsStyle="link"
  className="theme-tc-pie-chart-button tc-pie-chart-button"
  disabled={false}
  onClick={null}
  onMouseDown={null}
  role="button"
>
  <Translate(PieChartIcon)
    display="small"
    model={
      Array [
        Object {
          "color": "rio-grande",
          "percentage": 50,
        },
        Object {
          "color": "chestnut-rose",
          "percentage": 12,
        },
        Object {
          "color": "lightning-yellow",
          "percentage": 1,
        },
        Object {
          "color": "dove-gray",
          "percentage": 4,
        },
        Object {
          "color": "silver-chalice",
          "percentage": 3,
        },
      ]
    }
  />
</Button>
`;

exports[`PieChartButton snapshots render should render a PieChartButton with an overlay 1`] = `
<span>
  <OverlayTrigger
    defaultOverlayShown={false}
    overlay={
      <Popover
        bsClass="popover"
        id="id-popover"
        placement="right"
      >
        <div>
          I am an overlay
        </div>
      </Popover>
    }
    placement="bottom"
    rootClose={true}
    trigger="click"
  >
    <Button
      active={false}
      block={false}
      bsClass="btn"
      bsStyle="link"
      className="theme-tc-pie-chart-button tc-pie-chart-button"
      disabled={false}
      onClick={null}
      onMouseDown={null}
      role="button"
    >
      <Translate(PieChartIcon)
        display="medium"
        labelIndex={2}
        model={
          Array [
            Object {
              "color": "rio-grande",
              "percentage": 50,
            },
            Object {
              "color": "chestnut-rose",
              "percentage": 12,
            },
            Object {
              "color": "lightning-yellow",
              "percentage": 1,
            },
            Object {
              "color": "dove-gray",
              "percentage": 4,
            },
            Object {
              "color": "silver-chalice",
              "percentage": 3,
            },
          ]
        }
      />
    </Button>
  </OverlayTrigger>
</span>
`;

exports[`decorateWithOverlay should return the component wrapped 1`] = `
<span>
  <OverlayTrigger
    defaultOverlayShown={false}
    overlay={
      <Popover
        bsClass="popover"
        id="id-test"
        placement="right"
      >
        <div
          className="fake-overlay"
        />
      </Popover>
    }
    placement="top"
    rootClose={true}
    trigger="click"
  >
    <div
      className="fake-button-element"
    />
  </OverlayTrigger>
</span>
`;
